go/node/docker.go

```diff
@@ -0,0 +1,152 @@
+
+	err := d.startEdgelessDB()
+	if err != nil {
+		return err
+	}
+
+	err = d.startEnclave()
+	if err != nil {
+		return err
+	}
+
+	err = d.startHost()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *DockerNode) startHost() error {
+	cmd := []string{
+		"/home/obscuro/go-obscuro/go/host/main/main",
+		"-l1NodeHost", d.cfg.l1Host,
+		"-l1NodePort", fmt.Sprintf("%d", d.cfg.l1WSPort),
+		"-enclaveRPCAddress", fmt.Sprintf("%s:%d", d.cfg.nodeName+"-enclave", d.cfg.enclaveWSPort),
+		"-managementContractAddress", d.cfg.managementContractAddr,
+		"-privateKey", d.cfg.privateKey,
+		"-clientRPCHost", "0.0.0.0",
+		"-logPath", "sys_out",
+		"-logLevel", "4",
+		fmt.Sprintf("-isGenesis=%t", d.cfg.isGenesis), // boolean are a special case where the = is required
+		"-nodeType", d.cfg.nodeType,
+		"-profilerEnabled=false",
+		"-p2pPublicAddress", d.cfg.hostPublicP2PAddr,
+		"-p2pBindAddress", fmt.Sprintf("0.0.0.0:%d", d.cfg.hostP2PPort),
+		"-clientRPCPortHttp", fmt.Sprintf("%d", d.cfg.hostHTTPPort),
+		"-clientRPCPortWs", fmt.Sprintf("%d", d.cfg.hostWSPort),
+	}
+
+	exposedPorts := []int{
+		d.cfg.hostHTTPPort,
+		d.cfg.hostWSPort,
+		d.cfg.hostP2PPort,
+	}
+
+	_, err := docker.StartNewContainer(d.cfg.nodeName+"-host", d.cfg.hostImage, cmd, exposedPorts, nil, nil)
+
+	return err
+}
+
+func (d *DockerNode) startEnclave() error {
+	devices := map[string]string{}
+	exposedPorts := []int{}
+	envs := map[string]string{
+		"OE_SIMULATION": "1",
+	}
+
+	// default start of the enclave
+	cmd := []string{
+		"ego", "run", "/home/obscuro/go-obscuro/go/enclave/main/main",
+	}
+
+	if d.cfg.enclaveDebug {
+		cmd = []string{
+			"dlv",
+			"--listen=:2345",
+			"--headless=true",
+			"--log=true",
+			"--api-version=2",
+			"debug",
+			"/home/obscuro/go-obscuro/go/enclave/main",
+			"--",
+		}
+		exposedPorts = append(exposedPorts, 2345)
+	}
+
+	cmd = append(cmd,
+		"-hostID", d.cfg.hostID,
+		"-address", fmt.Sprintf("0.0.0.0:%d", d.cfg.enclaveWSPort), // todo review this 0.0.0.0 host bind
+		"-nodeType", d.cfg.nodeType,
+		"-managementContractAddress", d.cfg.managementContractAddr,
+		"-hostAddress", fmt.Sprintf("%s:%d", d.cfg.nodeName+"-host", d.cfg.hostP2PPort),
+		"-sequencerID", d.cfg.sequencerID,
+		"-messageBusAddress", d.cfg.messageBusContractAddress,
+		"-profilerEnabled=false",
+		"-useInMemoryDB=false",
+		"-logPath", "sys_out",
+		"-logLevel", "2",
+	)
+
+	if d.cfg.sgxEnabled {
+		devices["/dev/sgx_enclave"] = "/dev/sgx_enclave"
+		devices["/dev/sgx_provision"] = "/dev/sgx_provision"
+
+		envs["OE_SIMULATION"] = "0"
+		cmd = append(cmd,
+			"-edgelessDBHost", d.cfg.nodeName+"-edgelessdb",
+			"-willAttest=true",
+		)
+	} else {
+		cmd = append(cmd,
+			"-sqliteDBPath", "/data/sqlite.db",
+		)
+	}
+
+	_, err := docker.StartNewContainer(d.cfg.nodeName+"-enclave", d.cfg.enclaveImage, cmd, exposedPorts, envs, devices)
+	return err
+}
+
+func (d *DockerNode) startEdgelessDB() error {
+	if !d.cfg.sgxEnabled {
+		return nil
```

Comment:
Maybe worth a comment to say that on non-SGX hardware we use an in-mem sqlite database so no docker process required.

Reply:
Good point ! Added !